### PR TITLE
Evaluate setting `max_limit_for_ann_queries` during optimization pass

### PR DESCRIPTION
**Pull request by rschu1ze**
Addresses https://github.com/ClickHouse/ClickHouse/pull/71545#discussion_r1886920115

Note for reviewer: I first needed to do some cleanups and refactorings. Suggest to check the commits one-by-one. Commit [`ae849ff` (#73393)](https://github.com/ClickHouse/ClickHouse/pull/73393/commits/ae849ffc9e27ac03c708590df352fa2ce9aff7b5) contains the actual changes.

@novikd 

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit b08b1b14fc89ea46767c94e336440c954b59364a with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/73393/b08b1b14fc89ea46767c94e336440c954b59364a/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/954b9df6b66ea064d5de0c256ba6c6d946337604/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by rschu1ze** [comment]
ClickHouse Integration tests (aarch64) [2/6]:
- test_refreshable_mat_view_replicated/test.py::test_query_fail: #73430

File: src/Processors/QueryPlan/Optimizations/QueryPlanOptimizationSettings.h [hunk 1]
```diff
@@ -12,77 +12,52 @@ struct Settings;
...
     bool split_filter = true;
-
-    /// If merge-expressions optimization is enabled.
     bool merge_expressions = true;
-
-    /// If merge-filters optimization is enabled.
     bool merge_filters = true;
-
-    /// If filter push down optimization is enabled.
     bool filter_push_down = true;
```
Comment (jkartseva) [inline src/Processors/QueryPlan/Optimizations/QueryPlanOptimizationSettings.h:34]: I agree that the prior documentation was not informative.

File: src/Processors/QueryPlan/Optimizations/QueryPlanOptimizationSettings.h [hunk 2]
```diff
@@ -57,6 +57,7 @@ struct QueryPlanOptimizationSettings
     bool optimize_use_implicit_projections = false;
     bool force_use_projection = false;
     String force_projection_name;
+    size_t max_limit_for_ann_queries;
```
Comment (jkartseva) [inline src/Processors/QueryPlan/Optimizations/QueryPlanOptimizationSettings.h:60]: Should there be a default initialization value? For example, `0` which ultimately disables the optimization?

**Review by jkartseva — APPROVED** [review]
LGTM. Thanks for publishing refactoring changes in separate commits.

File: src/Processors/QueryPlan/Optimizations/QueryPlanOptimizationSettings.h (hunk 1 above)
Comment (rschu1ze) [reply src/Processors/QueryPlan/Optimizations/QueryPlanOptimizationSettings.h:34]: In particular, all these settings are documented exhaustively in Core/Settings.cpp so I was removing some redundancy.

File: src/Processors/QueryPlan/Optimizations/QueryPlanOptimizationSettings.h (hunk 2 above)
Comment (rschu1ze) [reply src/Processors/QueryPlan/Optimizations/QueryPlanOptimizationSettings.h:60]: Good point. It seems weird to me that the other settings above are initialized at all. As far as I see, the code always calls`QueryPlanOptimizationSettings::fromSettings` which makes the implicit initialization pointless.

Let me check if I can refactor this a little bit. I'd do this separately though as people are waiting for this PR :)

Comment (rschu1ze) [reply src/Processors/QueryPlan/Optimizations/QueryPlanOptimizationSettings.h]: Needed to merge from `master` anyways and fixed it ...

Comment (rschu1ze) [reply src/Processors/QueryPlan/Optimizations/QueryPlanOptimizationSettings.h]: --> https://github.com/ClickHouse/ClickHouse/pull/73693

**Pull request merged**